[c10d] Make test compatible for new pytest

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #136158

Temporary fix to the issue in https://github.com/pytorch/pytorch/issues/127517. 

Short-term fix following CPython: https://github.com/python/cpython/blob/51aefc5bf907ddffaaf083ded0de773adcdf08c8/Lib/unittest/case.py#L419-L426

Differential Revision: [D62878083](https://our.internmc.facebook.com/intern/diff/D62878083)

fduwjj committed:
[c10d] Make test compatible for new pytest

[ghstack-poisoned]

kwen2501 reviewed (commented):
Could the CI failure be related? I don't see an obvious reason though.

On the other hand, maybe the try-except method in cpython's example is safer?

fegin commented:
Can we also import the PR to test the internal CI?

fduwjj committed:
Update on "[c10d] Make test compatible for new pytest"


Temporary fix to the issue in https://github.com/pytorch/pytorch/issues/127517. 


[ghstack-poisoned]

fduwjj commented:
@fduwjj has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D62878083).

fduwjj commented:
@pytorchbot merge -i

fduwjj commented:
@pytorchbot merge -i
